File: minesweeper.go
// minesweeper.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	reset   = "\033[0m"
	blue    = "\033[34m"
	green   = "\033[32m"
	red     = "\033[31m"
	cyan    = "\033[36m"
	black   = "\033[30m"
	gray    = "\033[37m"
	yellow  = "\033[33m"
	white   = "\033[97m"
	bgDark  = "\033[100m"
	bgLight = "\033[47m"
)

var digitColors = map[rune]string{
	'1': blue,
	'2': green,
	'3': red,
	'4': blue,
	'5': red,
	'6': cyan,
	'7': black,
	'8': gray,
}

type Minesweeper struct {
	rows, cols, mines int
	board             [][]rune
	revealed          [][]bool
	flagged           [][]bool
	gameOver          bool
	won               bool
	firstMove         bool
	minePositions     map[string]bool
	startTime         time.Time
	elapsed           float64
	minesLeft         int
}

func NewMinesweeper(rows, cols, mines int) *Minesweeper {
	b := make([][]rune, rows)
	r := make([][]bool, rows)
	f := make([][]bool, rows)
	for i := 0; i < rows; i++ {
		b[i] = make([]rune, cols)
		r[i] = make([]bool, cols)
		f[i] = make([]bool, cols)
		for j := 0; j < cols; j++ {
			b[i][j] = ' '
		}
	}
	return &Minesweeper{
		rows:          rows,
		cols:          cols,
		mines:         mines,
		board:         b,
		revealed:      r,
		flagged:       f,
		firstMove:     true,
		minePositions: make(map[string]bool),
		minesLeft:     mines,
	}
}

func (m *Minesweeper) placeMines(firstRow, firstCol int) {
	safe := make(map[string]bool)
	for dr := -1; dr <= 1; dr++ {
		for dc := -1; dc <= 1; dc++ {
			nr, nc := firstRow+dr, firstCol+dc
			if nr >= 0 && nr < m.rows && nc >= 0 && nc < m.cols {
				safe[fmt.Sprintf("%d,%d", nr, nc)] = true
			}
		}
	}
	var candidates []string
	for r := 0; r < m.rows; r++ {
		for c := 0; c < m.cols; c++ {
			key := fmt.Sprintf("%d,%d", r, c)
			if !safe[key] {
				candidates = append(candidates, key)
			}
		}
	}
	// Перемешиваем
	rand.Shuffle(len(candidates), func(i, j int) { candidates[i], candidates[j] = candidates[j], candidates[i] })
	selected := candidates[:m.mines]
	for _, key := range selected {
		m.minePositions[key] = true
		var r, c int
		fmt.Sscanf(key, "%d,%d", &r, &c)
		m.board[r][c] = '*'
	}
}

func (m *Minesweeper) countNeighbors(r, c int) int {
	if m.board[r][c] == '*' {
		return -1
	}
	count := 0
	for dr := -1; dr <= 1; dr++ {
		for dc := -1; dc <= 1; dc++ {
			nr, nc := r+dr, c+dc
			if nr >= 0 && nr < m.rows && nc >= 0 && nc < m.cols && m.board[nr][nc] == '*' {
				count++
			}
		}
	}
	return count
}

func (m *Minesweeper) reveal(r, c int) {
	if m.gameOver || m.won {
		return
	}
	if r < 0 || r >= m.rows || c < 0 || c >= m.cols {
		return
	}
	if m.flagged[r][c] {
		return
	}
	if m.revealed[r][c] {
		return
	}
	if m.firstMove {
		m.firstMove = false
		m.placeMines(r, c)
		m.startTime = time.Now()
		for rr := 0; rr < m.rows; rr++ {
			for cc := 0; cc < m.cols; cc++ {
				if m.board[rr][cc] != '*' {
					m.board[rr][cc] = rune('0' + m.countNeighbors(rr, cc))
				}
			}
		}
	}
	if m.board[r][c] == '*' {
		m.gameOver = true
		m.revealAll()
		m.elapsed = time.Since(m.startTime).Seconds()
		return
	}
	m.revealed[r][c] = true
	if m.board[r][c] == '0' {
		for dr := -1; dr <= 1; dr++ {
			for dc := -1; dc <= 1; dc++ {
				nr, nc := r+dr, c+dc
				if nr >= 0 && nr < m.rows && nc >= 0 && nc < m.cols && !m.revealed[nr][nc] && !m.flagged[nr][nc] {
					m.reveal(nr, nc)
				}
			}
		}
	}
	m.checkWin()
}

func (m *Minesweeper) toggleFlag(r, c int) {
	if m.gameOver || m.won {
		return
	}
	if r < 0 || r >= m.rows || c < 0 || c >= m.cols {
		return
	}
	if m.revealed[r][c] {
		return
	}
	if m.flagged[r][c] {
		m.flagged[r][c] = false
		m.minesLeft++
	} else {
		m.flagged[r][c] = true
		m.minesLeft--
	}
}

func (m *Minesweeper) checkWin() {
	revealedCount := 0
	for r := 0; r < m.rows; r++ {
		for c := 0; c < m.cols; c++ {
			if m.revealed[r][c] {
				revealedCount++
			}
		}
	}
	if revealedCount == m.rows*m.cols-m.mines {
		m.won = true
		m.elapsed = time.Since(m.startTime).Seconds()
		m.gameOver = true
	}
}

func (m *Minesweeper) revealAll() {
	for r := 0; r < m.rows; r++ {
		for c := 0; c < m.cols; c++ {
			m.revealed[r][c] = true
		}
	}
}

func (m *Minesweeper) getDisplayChar(r, c int) string {
	if m.gameOver && !m.won {
		if m.board[r][c] == '*' {
			return "💣"
		}
		if m.revealed[r][c] {
			return string(m.board[r][c])
		}
		return " "
	}
	if m.revealed[r][c] {
		return string(m.board[r][c])
	}
	if m.flagged[r][c] {
		return "⚑"
	}
	return " "
}

func (m *Minesweeper) getColor(r, c int) string {
	if m.gameOver && !m.won {
		if m.board[r][c] == '*' {
			return red
		}
	}
	if !m.revealed[r][c] {
		if m.flagged[r][c] {
			return yellow
		}
		return white
	}
	ch := m.board[r][c]
	if ch >= '1' && ch <= '8' {
		if col, ok := digitColors[ch]; ok {
			return col
		}
	}
	return white
}

func (m *Minesweeper) render() {
	fmt.Print("  ")
	for c := 0; c < m.cols; c++ {
		fmt.Printf("%c ", 'a'+c)
	}
	fmt.Println()
	for r := 0; r < m.rows; r++ {
		fmt.Printf("%2d ", r+1)
		for c := 0; c < m.cols; c++ {
			ch := m.getDisplayChar(r, c)
			color := m.getColor(r, c)
			bg := bgDark
			if m.revealed[r][c] || (m.gameOver && !m.won) {
				bg = bgLight
			}
			if m.flagged[r][c] && !m.revealed[r][c] {
				bg = bgDark
			}
			fmt.Printf("%s%s%s%s ", bg, color, ch, reset)
		}
		fmt.Println()
	}
}

func (m *Minesweeper) play() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Добро пожаловать в Сапер!")
	fmt.Println("Вводите ход в формате: a1 (открыть) или f a1 (флаг)")
	fmt.Println("q - выход")
	for !m.gameOver {
		m.render()
		fmt.Printf("Мин осталось: %d\n", m.minesLeft)
		fmt.Print("Введите ход: ")
		if !scanner.Scan() {
			break
		}
		cmd := scanner.Text()
		if cmd == "q" {
			return
		}
		parts := strings.Fields(cmd)
		if len(parts) == 2 && parts[0] == "f" {
			coord := parts[1]
			if len(coord) < 2 {
				continue
			}
			col := int(coord[0] - 'a')
			row, _ := strconv.Atoi(coord[1:])
			m.toggleFlag(row-1, col)
		} else if len(parts) == 1 {
			coord := parts[0]
			if len(coord) < 2 {
				continue
			}
			col := int(coord[0] - 'a')
			row, _ := strconv.Atoi(coord[1:])
			m.reveal(row-1, col)
		}
	}
	m.render()
	if m.won {
		fmt.Printf("Поздравляем! Вы выиграли за %.1f секунд!\n", m.elapsed)
		// Сохраняем рекорд
		records := make(map[string]float64)
		if data, err := os.ReadFile("records.json"); err == nil {
			json.Unmarshal(data, &records)
		}
		key := fmt.Sprintf("%dx%d_%d", m.rows, m.cols, m.mines)
		if val, ok := records[key]; !ok || m.elapsed < val {
			records[key] = m.elapsed
			if data, err := json.MarshalIndent(records, "", "  "); err == nil {
				os.WriteFile("records.json", data, 0644)
			}
			fmt.Printf("Новый рекорд! %.1f сек.\n", m.elapsed)
		} else {
			fmt.Printf("Рекорд для этого уровня: %.1f сек.\n", val)
		}
	} else {
		fmt.Println("Вы проиграли. Попробуйте снова!")
	}
}

func chooseDifficulty() (int, int, int) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Выберите уровень сложности:")
	fmt.Println("1. Лёгкий (9x9, 10 мин)")
	fmt.Println("2. Средний (16x16, 40 мин)")
	fmt.Println("3. Сложный (30x16, 99 мин)")
	fmt.Print("Ваш выбор: ")
	scanner.Scan()
	choice := scanner.Text()
	switch choice {
	case "1":
		return 9, 9, 10
	case "2":
		return 16, 16, 40
	case "3":
		return 30, 16, 99
	default:
		fmt.Println("Некорректный выбор, установлен лёгкий.")
		return 9, 9, 10
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	rows, cols, mines := chooseDifficulty()
	game := NewMinesweeper(rows, cols, mines)
	game.play()
}

File: minesweeper.rs
// minesweeper.rs
use std::io::{self, Write, BufRead};
use std::fs;
use std::collections::HashMap;
use std::time::{Instant, Duration};
use rand::Rng;
use rand::seq::SliceRandom;

const RESET: &str = "\x1b[0m";
const BLUE: &str = "\x1b[34m";
const GREEN: &str = "\x1b[32m";
const RED: &str = "\x1b[31m";
const CYAN: &str = "\x1b[36m";
const BLACK: &str = "\x1b[30m";
const GRAY: &str = "\x1b[37m";
const YELLOW: &str = "\x1b[33m";
const WHITE: &str = "\x1b[97m";
const BG_DARK: &str = "\x1b[100m";
const BG_LIGHT: &str = "\x1b[47m";

fn digit_color(ch: char) -> &'static str {
    match ch {
        '1' => BLUE,
        '2' => GREEN,
        '3' => RED,
        '4' => BLUE,
        '5' => RED,
        '6' => CYAN,
        '7' => BLACK,
        '8' => GRAY,
        _ => WHITE,
    }
}

struct Minesweeper {
    rows: usize,
    cols: usize,
    mines: usize,
    board: Vec<Vec<char>>,
    revealed: Vec<Vec<bool>>,
    flagged: Vec<Vec<bool>>,
    game_over: bool,
    won: bool,
    first_move: bool,
    mine_positions: Vec<(usize, usize)>,
    start_time: Option<Instant>,
    elapsed: f64,
    mines_left: i32,
}

impl Minesweeper {
    fn new(rows: usize, cols: usize, mines: usize) -> Self {
        Minesweeper {
            rows,
            cols,
            mines,
            board: vec![vec![' '; cols]; rows],
            revealed: vec![vec![false; cols]; rows],
            flagged: vec![vec![false; cols]; rows],
            game_over: false,
            won: false,
            first_move: true,
            mine_positions: Vec::new(),
            start_time: None,
            elapsed: 0.0,
            mines_left: mines as i32,
        }
    }

    fn place_mines(&mut self, first_row: usize, first_col: usize) {
        let mut safe = std::collections::HashSet::new();
        for dr in -1..=1 {
            for dc in -1..=1 {
                let nr = first_row as i32 + dr;
                let nc = first_col as i32 + dc;
                if nr >= 0 && nr < self.rows as i32 && nc >= 0 && nc < self.cols as i32 {
                    safe.insert((nr as usize, nc as usize));
                }
            }
        }
        let mut candidates = Vec::new();
        for r in 0..self.rows {
            for c in 0..self.cols {
                if !safe.contains(&(r, c)) {
                    candidates.push((r, c));
                }
            }
        }
        let mut rng = rand::thread_rng();
        candidates.shuffle(&mut rng);
        self.mine_positions = candidates[..self.mines].to_vec();
        for &(r, c) in &self.mine_positions {
            self.board[r][c] = '*';
        }
    }

    fn count_neighbors(&self, r: usize, c: usize) -> i32 {
        if self.board[r][c] == '*' { return -1; }
        let mut count = 0;
        for dr in -1..=1 {
            for dc in -1..=1 {
                let nr = r as i32 + dr;
                let nc = c as i32 + dc;
                if nr >= 0 && nr < self.rows as i32 && nc >= 0 && nc < self.cols as i32 {
                    if self.board[nr as usize][nc as usize] == '*' {
                        count += 1;
                    }
                }
            }
        }
        count
    }

    fn reveal(&mut self, r: usize, c: usize) {
        if self.game_over || self.won { return; }
        if r >= self.rows || c >= self.cols { return; }
        if self.flagged[r][c] { return; }
        if self.revealed[r][c] { return; }
        if self.first_move {
            self.first_move = false;
            self.place_mines(r, c);
            self.start_time = Some(Instant::now());
            for rr in 0..self.rows {
                for cc in 0..self.cols {
                    if self.board[rr][cc] != '*' {
                        let cnt = self.count_neighbors(rr, cc);
                        self.board[rr][cc] = if cnt == 0 { ' ' } else { char::from_digit(cnt as u32, 10).unwrap() };
                    }
                }
            }
        }
        if self.board[r][c] == '*' {
            self.game_over = true;
            self.reveal_all();
            if let Some(st) = self.start_time {
                self.elapsed = st.elapsed().as_secs_f64();
            }
            return;
        }
        self.revealed[r][c] = true;
        if self.board[r][c] == ' ' { // пустая клетка (0)
            for dr in -1..=1 {
                for dc in -1..=1 {
                    let nr = r as i32 + dr;
                    let nc = c as i32 + dc;
                    if nr >= 0 && nr < self.rows as i32 && nc >= 0 && nc < self.cols as i32 {
                        let (nr, nc) = (nr as usize, nc as usize);
                        if !self.revealed[nr][nc] && !self.flagged[nr][nc] {
                            self.reveal(nr, nc);
                        }
                    }
                }
            }
        }
        self.check_win();
    }

    fn toggle_flag(&mut self, r: usize, c: usize) {
        if self.game_over || self.won { return; }
        if r >= self.rows || c >= self.cols { return; }
        if self.revealed[r][c] { return; }
        if self.flagged[r][c] {
            self.flagged[r][c] = false;
            self.mines_left += 1;
        } else {
            self.flagged[r][c] = true;
            self.mines_left -= 1;
        }
    }

    fn check_win(&mut self) {
        let mut revealed_count = 0;
        for r in 0..self.rows {
            for c in 0..self.cols {
                if self.revealed[r][c] { revealed_count += 1; }
            }
        }
        if revealed_count == self.rows * self.cols - self.mines {
            self.won = true;
            if let Some(st) = self.start_time {
                self.elapsed = st.elapsed().as_secs_f64();
            }
            self.game_over = true;
        }
    }

    fn reveal_all(&mut self) {
        for r in 0..self.rows {
            for c in 0..self.cols {
                self.revealed[r][c] = true;
            }
        }
    }

    fn get_display_char(&self, r: usize, c: usize) -> char {
        if self.game_over && !self.won {
            if self.board[r][c] == '*' { return '💣'; }
            if self.revealed[r][c] { return self.board[r][c]; }
            return ' ';
        }
        if self.revealed[r][c] { return self.board[r][c]; }
        if self.flagged[r][c] { return '⚑'; }
        ' '
    }

    fn get_color(&self, r: usize, c: usize) -> &'static str {
        if self.game_over && !self.won {
            if self.board[r][c] == '*' { return RED; }
        }
        if !self.revealed[r][c] {
            if self.flagged[r][c] { return YELLOW; }
            return WHITE;
        }
        let ch = self.board[r][c];
        if ch.is_ascii_digit() {
            digit_color(ch)
        } else {
            WHITE
        }
    }

    fn render(&self) {
        print!("  ");
        for c in 0..self.cols {
            print!("{} ", (b'a' + c as u8) as char);
        }
        println!();
        for r in 0..self.rows {
            print!("{:2} ", r+1);
            for c in 0..self.cols {
                let ch = self.get_display_char(r, c);
                let color = self.get_color(r, c);
                let bg = if self.revealed[r][c] || (self.game_over && !self.won) { BG_LIGHT } else { BG_DARK };
                print!("{}{}{}{} ", bg, color, ch, RESET);
            }
            println!();
        }
    }

    fn play(&mut self) {
        let stdin = io::stdin();
        let mut input = String::new();
        println!("Добро пожаловать в Сапер!");
        println!("Вводите ход в формате: a1 (открыть) или f a1 (флаг)");
        println!("q - выход");
        while !self.game_over {
            self.render();
            println!("Мин осталось: {}", self.mines_left);
            print!("Введите ход: ");
            io::stdout().flush().unwrap();
            input.clear();
            stdin.read_line(&mut input).unwrap();
            let cmd = input.trim();
            if cmd == "q" { break; }
            let parts: Vec<&str> = cmd.split_whitespace().collect();
            if parts.len() == 2 && parts[0] == "f" {
                let coord = parts[1];
                if coord.len() < 2 { continue; }
                let col = coord.chars().next().unwrap() as usize - 'a' as usize;
                let row = coord[1..].parse::<usize>().unwrap_or(1) - 1;
                self.toggle_flag(row, col);
            } else if parts.len() == 1 {
                let coord = parts[0];
                if coord.len() < 2 { continue; }
                let col = coord.chars().next().unwrap() as usize - 'a' as usize;
                let row = coord[1..].parse::<usize>().unwrap_or(1) - 1;
                self.reveal(row, col);
            }
        }
        self.render();
        if self.won {
            println!("Поздравляем! Вы выиграли за {:.1} секунд!", self.elapsed);
            // Чтение рекордов
            let key = format!("{}x{}_{}", self.rows, self.cols, self.mines);
            let mut records: HashMap<String, f64> = HashMap::new();
            if let Ok(data) = fs::read_to_string("records.json") {
                if let Ok(parsed) = serde_json::from_str(&data) {
                    records = parsed;
                }
            }
            let old = records.get(&key).copied();
            if old.is_none() || self.elapsed < old.unwrap() {
                records.insert(key, self.elapsed);
                if let Ok(json) = serde_json::to_string_pretty(&records) {
                    let _ = fs::write("records.json", json);
                }
                println!("Новый рекорд! {:.1} сек.", self.elapsed);
            } else {
                println!("Рекорд для этого уровня: {:.1} сек.", old.unwrap());
            }
        } else {
            println!("Вы проиграли. Попробуйте снова!");
        }
    }
}

fn choose_difficulty() -> (usize, usize, usize) {
    println!("Выберите уровень сложности:");
    println!("1. Лёгкий (9x9, 10 мин)");
    println!("2. Средний (16x16, 40 мин)");
    println!("3. Сложный (30x16, 99 мин)");
    print!("Ваш выбор: ");
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    match input.trim() {
        "1" => (9, 9, 10),
        "2" => (16, 16, 40),
        "3" => (30, 16, 99),
        _ => {
            println!("Некорректный выбор, установлен лёгкий.");
            (9, 9, 10)
        }
    }
}

fn main() {
    let (rows, cols, mines) = choose_difficulty();
    let mut game = Minesweeper::new(rows, cols, mines);
    game.play();
}
